### ISSUE TITLE
Form type "Sonata\CoreBundle\Form\Type\TranslatableChoiceType" is deprecated

### ISSUE THREAD

**Issue by gnutix** [sonata-project/SonataCoreBundle#532]
#### Sonata packages

```
$ composer show --latest 'sonata-project/*'
sonata-project/admin-bundle              3.33.0 3.33.0 The missing Symfony Admin Generator
sonata-project/block-bundle              3.12.1 3.12.1 Symfony SonataBlockBundle
sonata-project/cache                     1.0.7  1.0.7  Cache library
sonata-project/core-bundle               3.9.1  3.9.1  Symfony SonataCoreBundle
sonata-project/datagrid-bundle           2.3.1  2.3.1  Symfony SonataDatagridBundle
sonata-project/doctrine-orm-admin-bundle 3.4.2  3.4.2  Symfony Sonata / Integrate Doctrine ORM into the SonataAdminBundle
sonata-project/exporter                  1.8.0  1.8.0  Lightweight Exporter library
```

#### Symfony packages

```
$ composer show --latest 'symfony/*'
symfony/assetic-bundle      v2.8.2 v2.8.2 Integrates Assetic into Symfony2
symfony/monolog-bundle      v3.2.0 v3.2.0 Symfony MonologBundle
symfony/phpunit-bridge      v4.0.7 v4.0.7 Symfony PHPUnit Bridge
symfony/polyfill-apcu       v1.7.0 v1.7.0 Symfony polyfill backporting apcu_* functions to lower PHP versions
symfony/polyfill-intl-icu   v1.7.0 v1.7.0 Symfony polyfill for intl's ICU-related data and classes
symfony/polyfill-mbstring   v1.7.0 v1.7.0 Symfony polyfill for the Mbstring extension
symfony/polyfill-php56      v1.7.0 v1.7.0 Symfony polyfill backporting some PHP 5.6+ features to lower PHP versions
symfony/polyfill-php70      v1.7.0 v1.7.0 Symfony polyfill backporting some PHP 7.0+ features to lower PHP versions
symfony/polyfill-util       v1.7.0 v1.7.0 Symfony utilities for portability of PHP codes
symfony/security-acl        v3.0.1 v3.0.1 Symfony Security Component - ACL (Access Control List)
symfony/swiftmailer-bundle  v2.6.7 v3.0.3 Symfony SwiftmailerBundle
symfony/symfony             v3.4.7 v3.4.7 The Symfony PHP framework
symfony/web-profiler-bundle v3.4.6 v3.4.7 Symfony WebProfilerBundle
```

#### PHP version

```
$ php -v
PHP 5.6.33-0+deb8u1 (cli) (built: Jan  5 2018 15:46:26) 
Copyright (c) 1997-2016 The PHP Group
Zend Engine v2.6.0, Copyright (c) 1998-2016 Zend Technologies
    with Zend OPcache v7.0.6-dev, Copyright (c) 1999-2016, by Zend Technologies
    with blackfire v1.18.0~linux-x64-non_zts56, https://blackfire.io, by SensioLabs
```

## Subject

I'm currently in the process of upgrading an application from Symfony 2.3 to 3.4. When loading the homepage, I get this deprecation notice in the profiler : 

> User Deprecated: Form type "Sonata\CoreBundle\Form\Type\TranslatableChoiceType" is deprecated since SonataCoreBundle 2.2.0 and will be removed in 4.0. Use form type "Symfony\Component\Form\Extension\Core\Type\ChoiceType" with "translation_domain" option instead.

However, I can't find any direct usage of the `Sonata\CoreBundle\Form\Type\TranslatableChoiceType` class. I've upgraded all the `FormBuilder->add()` calls to use the new `ClassType::class` syntax and haven't seen it. The only occurence of `TranslatableChoiceType` comes from `vendor/sonata/core-bundle/src` (the service definition, adding the class to the DI compilation, the form type itself, registering the form type mapping `sonata_type_translatable_choice`...). So it's not "used" (as far as I understand). Any idea how to get rid of the warning ?

**Comment by gnutix** [comment]
I've just seen that the call to `@trigger_error` is done at the top of the file... which explains that the deprecation message appears even though the class is not really "used". So I can't get rid of the deprecation message until Sonata get rid of the class completely, right ? ... Doesn't feel great.

**Comment by kunicmarko20** [comment]
Yes, if it is on the top it is triggered on autoload, that is why you are getting it even if you are not using it.

**Comment by gnutix** [comment]
Wouldn't it be better to have it in a method like `configureOptions`, which will be called if the class is actually used ?

**Comment by greg0ire** [comment]
If you want to know when the class is used, you can replace the `trigger_error` with an exception, this should give you a stack trace.

**Comment by kiler129** [comment]
I also stepped into that. When container is built during tests I'm getting:

```
1x: The Sonata\CoreBundle\Form\Type\ColorType class is deprecated since version 3.10 and will be removed in 4.0. Use Symfony\Component\Form\Extension\Core\Type\ColorType instead.
    1x in AddressTest::__construct from ***\***\Tests\Entity
```

I added exception to get stack trace and that's what I've got:
```
#0 /mnt/[REDACTED]/bin/.phpunit/phpunit-6.5/vendor/composer/ClassLoader.php(444): include()
#1 /mnt/[REDACTED]/bin/.phpunit/phpunit-6.5/vendor/composer/ClassLoader.php(322): Composer\Autoload\includeFile('/mnt/sdb1/publi...')
#2 [internal function]: Composer\Autoload\ClassLoader->loadClass('Sonata\\CoreBund...')
#3 [internal function]: spl_autoload_call('Sonata\\CoreBund...')
#4 /mnt/[REDACTED]/vendor/symfony/config/Resource/ClassExistenceResource.php(76): class_exists('Sonata\\CoreBund...')
#5 /mnt/[REDACTED]/vendor/symfony/dependency-injection/ContainerBuilder.php(350): Symfony\Component\Config\Resource\ClassExistenceResource->isFresh(0)
#6 /mnt/[REDACTED]/vendor/symfony/dependency-injection/Compiler/AutowirePass.php(311): Symfony\Component\DependencyInjection\ContainerBuilder->getReflectionCl in /mnt/[REDACTED]/vendor/sonata-project/core-bundle/src/Form/Type/ColorType.php on line 25
```

In summary the deprecation cannot be triggered when file loads. There are few other examples of the same problem:
 - https://github.com/sonata-project/SonataCoreBundle/blob/2a86fb17d6028ea69620ece1caad3ac36aed1a14/src/Component/NativeSlugify.php#L14-L17
 - https://github.com/sonata-project/SonataCoreBundle/blob/352275323c28f921c57312d5fb609b693a8bc8e0/src/Form/Type/ColorType.php#L18-L22
 - https://github.com/sonata-project/SonataCoreBundle/blob/2a86fb17d6028ea69620ece1caad3ac36aed1a14/src/Form/Type/ColorSelectorType.php#L20-L24
 - https://github.com/sonata-project/SonataCoreBundle/blob/2a86fb17d6028ea69620ece1caad3ac36aed1a14/src/Form/Type/TranslatableChoiceType.php#L22-L30
 - https://github.com/sonata-project/SonataCoreBundle/blob/efa2b90711c8eb648eb576703bf5cbc35bcc0da2/src/Exporter/Exporter.php#L21-L25
 - https://github.com/sonata-project/SonataCoreBundle/blob/352275323c28f921c57312d5fb609b693a8bc8e0/src/Color/Colors.php#L14-L17

**Comment by greg0ire** [comment]
So it has to do with autowiring :thinking: ?

**Comment by kiler129** [comment]
@greg0ire The assumption of this deprecation is nothing is going to include the class until it's used, which in case of autowire is not true. Symfony basically follows PSR (and reasonable assumptions) that loading class file will not cause side effects.

`AutowirePass` indeed scans all services which are autowire candidates and with 4.x the process of container intelligence got even more extensive.

**Comment by greg0ire** [comment]
Then why do they do it themselves: https://github.com/symfony/symfony/blob/43f8081d60abc5719518e5ff68b5d280f6e4bf7b/src/Symfony/Bundle/SecurityBundle/SecurityUserValueResolver.php

Maybe the issue is not doing that here, maybe the issue is that it is considered an autowiring candidate?

**Comment by kiler129** [comment]
As per discussion on slack `ColorType` is not marked as deprecated in services config. So basically changing the definition to:
```xml
        <service id="sonata.core.form.type.color" class="Sonata\CoreBundle\Form\Type\ColorType">
            <deprecated>The "%service_id%" service is deprecated since 3.5 and will be removed in 4.0.</deprecated>
            <tag name="form.type" alias="sonata_type_color"/>
        </service>
```

Will get rid of the `ColorType` deprecation, but it doesn't explain `TranslatableChoiceType` problem.

**Comment by lukepass** [comment]
Any news on this issue? I am using Symfony 3.4 and I got this warning.

Thanks.

**Comment by tifabien** [comment]
I think #651 will fix this issue.

**Comment by core23** [comment]
Should be fixed in #667
